gnome-image-installer/pages/diskimage/gis-diskimage-page.c

```diff
@@ -242,16 +267,95 @@ static void add_image(GtkListStore *store, gchar *image)
   g_object_unref(fi);
 }
 
+static gboolean
+file_exists (
+    const gchar *path,
+    GError     **error)
+{
+  g_autoptr(GFile) file = g_file_new_for_path (path);
+
+  if (g_file_query_exists (file, NULL))
+    return TRUE;
+
+  g_set_error (error, GIS_IMAGE_ERROR, 0, "%s doesn't exist", path);
+  return FALSE;
+}
+
+/* live USB sticks have an unpacked disk image named endless.img, which for
+ * various reasons is invisible in directory listings. We can determine the
+ * name that the image "should" have by reading /endless/live, and find the
+ * corresponding signature file.
+ */
+static gboolean
+gis_diskimage_page_add_live_image (
+    GtkListStore *store,
+    gchar        *path,
+    GError      **error)
+{
+  g_autofree gchar *endless_img_path = g_build_path (
+      "/", path, "endless", "endless.img", NULL);
+  g_autofree gchar *live_flag_path = g_build_path (
+      "/", path, "endless", "live", NULL);
+  g_autofree gchar *live_flag_contents = NULL;
+  g_autofree gchar *live_sig_basename = NULL;
+  g_autofree gchar *live_sig = NULL;
+
+  if (!file_exists (endless_img_path, error))
+    {
+      return FALSE;
+    }
+
+  if (!g_file_get_contents (live_flag_path, &live_flag_contents, NULL,
+        error))
+    {
+      g_prefix_error (error, "Couldn't read %s: ", live_flag_path);
+      return FALSE;
+    }
+
+  /* live_flag_contents contains the name that 'endless.img' would have had;
+   * so we should be able to find its signature at ${live_flag_contents}.asc
+   */
+  g_strstrip (live_flag_contents);
+  live_sig_basename = g_strdup_printf ("%s.%s", live_flag_contents, "asc");
+  live_sig = g_build_path ("/", path, "endless", live_sig_basename, NULL);
+
+  if (!file_exists (live_sig, error))
+    {
+      return FALSE;
+    }
+
+  if (file_exists (live_device_path, NULL))
+    {
+      add_image (store, live_device_path, live_sig);
+    }
+  else
+    {
+      g_print ("can't find image device %s; will use %s directly\n",
+               live_device_path, endless_img_path);
+      add_image (store, endless_img_path, live_sig);
+    }
+
+  return TRUE;
+}
+
 static void
 gis_diskimage_page_populate_model(GisPage *page, gchar *path)
 {
   GError *error = NULL;
+  gchar *imgpath = NULL;
   gchar *file = NULL;
   gchar *ufile = NULL;
   GtkListStore *store = OBJ(GtkListStore*, "image_store");
-  GDir *dir = g_dir_open (path, 0, &error);
+  GDir *dir;
   GtkTreeIter iter;
+  gboolean is_live = gis_store_is_live_install ();
+
+  if (is_live)
+    imgpath = g_build_path ("/", path, "endless", NULL);
```

Comment:
I guess this is maybe a little academic, but perhaps to "keep the meat from the flies" when we are acting on separate .img.*z{,.asc} files with the installer running under a live system, rather than reformatting the host with the running live system, maybe we should still look in the root folder and not do this "endless" subdir check any more?